File: packages/constants/src/worklog.ts
/**
 * Copyright (c) 2023-present Plane Software, Inc. and contributors
 * SPDX-License-Identifier: AGPL-3.0-only
 * See the LICENSE file for details.
 */

/**
 * Format minutes to a human-readable display string.
 * e.g. 90 → "1h 30m", 60 → "1h", 25 → "25m", 0 → "0m"
 */
export const formatMinutesToDisplay = (minutes: number): string => {
  const h = Math.floor(minutes / 60);
  const m = minutes % 60;
  if (h === 0) return `${m}m`;
  if (m === 0) return `${h}h`;
  return `${h}h ${m}m`;
};

/**
 * Convert hours + minutes inputs to total minutes.
 */
export const parseDisplayToMinutes = (hours: number, minutes: number): number => hours * 60 + minutes;

File: apps/api/plane/utils/worklog.py
# Copyright (c) 2023-present Plane Software, Inc. and contributors
# SPDX-License-Identifier: AGPL-3.0-only
# See the LICENSE file for details.

"""Canonical worklog duration helpers.

Duration is stored as a positive integer number of seconds. Human-readable
strings such as ``1h 30m`` are never persisted.
"""

WORKLOG_DURATION_MIN_SECONDS = 1
# 10,000 hours — a single entry larger than this is treated as overflow.
WORKLOG_DURATION_MAX_SECONDS = 10_000 * 3600


def validate_worklog_duration(value):
    """Return a validated integer duration in seconds or raise ValueError."""
    if isinstance(value, bool) or value is None:
        raise ValueError("Duration must be a positive integer number of seconds.")
    if isinstance(value, str):
        stripped = value.strip()
        if not stripped.isdigit():
            raise ValueError("Duration must be a positive integer number of seconds.")
        value = int(stripped)
    if isinstance(value, float):
        if not value.is_integer():
            raise ValueError("Duration must be a whole number of seconds.")
        value = int(value)
    if not isinstance(value, int):
        raise ValueError("Duration must be a positive integer number of seconds.")
    if value < WORKLOG_DURATION_MIN_SECONDS:
        raise ValueError("Duration must be greater than zero.")
    if value > WORKLOG_DURATION_MAX_SECONDS:
        raise ValueError("Duration exceeds the maximum allowed value.")
    return value
